server/src/lib/model/db/migrations/20200603213417-add-demographics-table.ts

```diff
@@ -0,0 +1,79 @@
+export const up = async function (db: any): Promise<any> {
+  return db.runSql(
+    `
+      CREATE TABLE ages (
+        id BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY,
+        age VARCHAR(255) NOT NULL
+      );
+
+      CREATE TABLE sexes (
+        id BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY,
+        sex VARCHAR(255) NOT NULL
+      );
+
+      CREATE TABLE demographics (
+        id BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY,
+        client_id CHAR(36) NOT NULL,
+        age_id BIGINT(20),
+        sex_id BIGINT(20),
+        updated_at DATETIME DEFAULT now(),
+        FOREIGN KEY (client_id) REFERENCES user_clients (client_id),
+        FOREIGN KEY (age_id) REFERENCES ages (id),
+        FOREIGN KEY (sex_id) REFERENCES sexes (id),
+        UNIQUE (client_id, age_id, sex_id)
+      );
+
+      CREATE TABLE clip_demographics (
+        clip_id BIGINT(20) UNSIGNED NOT NULL PRIMARY KEY,
+        demographic_id BIGINT(20) UNSIGNED NOT NULL,
+        FOREIGN KEY (clip_id) REFERENCES clips (id),
+        FOREIGN KEY (demographic_id) REFERENCES demographics (id)
+      );
+
+      ALTER TABLE user_clients
+        CHANGE COLUMN age deprecated_age VARCHAR(255),
+        CHANGE COLUMN gender deprecated_gender VARCHAR(255);
+
+      INSERT INTO ages (age) VALUES
+        ('teens'),
+        ('twenties'),
+        ('thirties'),
+        ('fourties'),
+        ('fifties'),
+        ('sixties'),
+        ('seventies'),
+        ('eighties'),
+        ('nineties');
+
+      /* Sex and gender exist across multiple spectra which are not adequately
+         represented in our current database schema.
+
+         Including demographic data with our dataset is necessary to ensure its
+         utility to researchers, and helps us track the fullness and inclusion
+         of our collection process. We will ALWAYS keep this field optional. We
+         only store records for users who opt in, and will always (at very
+         least) maintain an "Other" option.
+      */
+      INSERT INTO sexes (sex) VALUES ('female'), ('male'), ('other');
+
+      INSERT INTO demographics (client_id, age_id, sex_id)  (
+        SELECT user_clients.client_id, ages.id, sexes.id
+        FROM user_clients
+        LEFT JOIN ages ON ages.age = user_clients.deprecated_age
+        LEFT JOIN sexes ON sexes.sex = user_clients.deprecated_gender
```

Comment:
Out of an abundance of caution, what do you think about wrapping the deprecated fields in `LOWER()` and `TRIM()`?

Reply:
Done! And I switched:

```
        WHERE user_clients.deprecated_age IS NOT NULL
           OR user_clients.deprecated_gender IS NOT NULL
```

with:

```
        WHERE ages.age IS NOT NULL
           OR genders.gender IS NOT NULL
```

to ensure that accounts with invalid and unsalvageable demographic entries don't make it into this table.

---

(I also fixed all invalid ages / genders in our DB, but just in case something breaks before this hits prod)